### DISCUSSION TITLE
Bindings not working

### DISCUSSION THREAD

**Discussion by logiclrd** [AvaloniaUI/Avalonia#15902] [Q&A] [answered]
I have a new Avalonia UI project I'm working on. I have fairly extensive experience with WPF on Windows but this is the first time I've worked with Avalonia UI. I'm finding that bindings aren't working and I'm not sure what I'm doing wrong.

For instance, I have a command in `App.axaml.cs`:
```
	public App()
	{
		ShowWindowCommand = ReactiveCommand.Create(ShowWindow);
		...
	}

	public ICommand ShowWindowCommand { get; }

	...

	public MainWindow? MainWindow { get; set; }

	void ShowWindow()
	{
		MainWindow ??= new MainWindow();
		MainWindow.Show();
	}
```

Then in `App.axaml`, I'm defining a `TrayIcon` with a menu:
```
   <TrayIcon.Icons>
        <TrayIcons>
            <TrayIcon Icon="DeltaQ.RTB.ico" ToolTipText="DeltaQ Real-Time Backup">
                <TrayIcon.Menu>
                    <NativeMenu>
                        <NativeMenuItem Header="_Show Status Window" Command="{Binding ShowWindowCommand}" />
                        ...
 ```

The tray icon appears, and when clicked on, the menu shows up, but when I click on "Show Status Window", nothing happens. `ShowWindow` isn't called.

I have another class, a `UserControl` in charge of displaying details from an `UploadStatus` object:
```
    <UserControl.Resources>
        <converters:UploadStatusSpeedConverter x:Key="UploadStatusSpeedConverter" />
        <converters:UploadStatusProgressConverter x:Key="UploadStatusProgressConverter" />
        <converters:UploadStatusProgressTextConverter x:Key="UploadStatusProgressTextConverter" />
    </UserControl.Resources>

    <Grid>
        ...
        <Label Grid.Column="0" Content="{Binding UploadStatus, Converter={StaticResource UploadStatusSpeedConverter}}" />
        <ProgressBar Grid.Column="1" Minimum="0" Maximum="1" Value="{Binding UploadStatus, Converter={StaticResource UploadStatusProgressConverter}}" />
        <Label Grid.Column="2" Content="{Binding UploadStatus, Converter={StaticResource UploadStatusProgressTextConverter}}" />
        <Label Grid.Column="3" Content="{Binding UploadStatus, Path=Path}" />
    </Grid>
```

In the code-behind file, the `UploadStatus` properly is defined like this:
```
		public static readonly StyledProperty<UploadStatus> UploadStatusProperty =
			AvaloniaProperty.Register<UploadStatusControl, UploadStatus>(nameof(UploadStatus));

		public UploadStatus UploadStatus
		{
			get => GetValue(UploadStatusProperty);
			set => SetValue(UploadStatusProperty, value);
		}
```

But, setting `UploadStatus` to a value does not trigger any of the data binding. The control just remains blank. When I put a breakpoint into `Convert` in any of the `IValueConverter` classes, it is not hit.

At some point in the past, at least some of the data binding _was_ working. I created the control a couple of days ago and, at the time, did a test of it with dummy data to get it into the state I wanted. Then I added a bunch of other stuff to the project and wired up the logic that obtains actual status to be displayed, and when I wired that up to the control, data binding wasn't working. I've been poking at it for an hour now and can't figure out what the problem is.

Does anyone know what to look at? Am I doing something obviously wrong? How do I debug this? :-)

**Comment by stevemonaco** [comment, marked as answer]
`<NativeMenuItem Header="_Show Status Window" Command="{Binding ShowWindowCommand}" />`

Bindings implicitly go through the `DataContext` (usually a VM). You seem to have a method on the View (App) and it's named `ShowWindow` anyways.

Similarly for the second one. You need to bind to the control, not a ViewModel. So you can do `{Binding $parent[UserControl].UploadStatus}` or give it a name and `{Binding #myControl.UploadStatus}`. Add the converter afterwards.

Enable compiled bindings if you aren't using them. You probably aren't, else those should be caught at compile-time.

**Reply by logiclrd** [reply]
I see what you're saying and will see if I can figure it out. I just want to point out, though, that the method `ShowWindow` is only there to be an action for the command `ShowWindowCommand`, and it is the latter to which the binding is referring.

Hmm, I wonder if it would work to throw this into the constructor:
```
    DataContext = this;
```

**Reply by logiclrd** [reply]
In fact, I think I've written that exact line in WPF code in the past. :-P I should have remembered!

Yes, it seems to work!

**Reply by stevemonaco** [reply]
Yes, that works if you're using the View as the source of your binding state. Most people are using MVVM, but I sometimes use the View itself like that for quick-and-dirty demos.
